integration/benchmark/main.go

```diff
@@ -0,0 +1,234 @@
+	"github.com/containerd/containerd/containers"
+	"github.com/containerd/containerd/log/logtest"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/oci"
+	"github.com/containerd/containerd/platforms"
+	"github.com/montanaflynn/stats"
+)
+
+var (
+	address        = "/run/containerd/containerd.sock"
+	testImage      = "ghcr.io/containerd/busybox:1.28"
+	testNamespace  = "testing"
+	benchmarkFuncs = [...]func(*testing.B){BenchmarkRandom, BenchmarkGetImage, BenchmarkContainerCreate}
+)
+
+// https://stackoverflow.com/a/70535822/13675
+func GetFunctionName(function interface{}) string {
+	strs := strings.Split((runtime.FuncForPC(reflect.ValueOf(function).Pointer()).Name()), ".")
+	return strs[len(strs)-1]
+}
+
+func main() {
+	commit := os.Args[1]
+	testNum := 100 // less than 4 can produce fatal errors in the stats calculations
+	testing.Init()
+	// disable the automatic iteration count system
+	flag.Set("test.benchtime", "1x")
+	flag.Parse()
+
+	commitResults := BenchmarkCommitResults{
+		Commit:  commit,
+		Results: make(map[string]BenchmarkResult)}
+
+	for _, benchmarkFunc := range benchmarkFuncs {
+		benchmarkResult := BenchmarkResult{
+			TestName:      GetFunctionName(benchmarkFunc),
+			NumberOfTests: testNum,
+			testFunction:  benchmarkFunc}
+
+		for i := 0; i < benchmarkResult.NumberOfTests; i++ {
+			// fmt.Printf("TestName: %s\n", benchmarkResult.TestName)
+			res := testing.Benchmark(benchmarkResult.testFunction)
+			// fmt.Printf("%d: %f\n", i, res.T.Seconds())
+			benchmarkResult.addResult(res)
+		}
+		benchmarkResult.updateStats()
+
+		commitResults.Results[benchmarkResult.TestName] = benchmarkResult
+	}
+
+	json, _ := json.MarshalIndent(commitResults, "", " ")
+	fmt.Print(string(json[:]))
+}
+
+type BenchmarkCommitResults struct {
+	Commit  string                     `json:"commit"`
+	Results map[string]BenchmarkResult `json:"benchmarkResults"`
+}
+
+type BenchmarkResult struct {
+	TestName      string `json:"testName"`
+	NumberOfTests int    `json:"numberOfTests"`
+	testFunction  func(*testing.B)
+	TestsRun      int       `json:"testsRun"`
+	testTimes     []float64 `json:"-"`
+	StdDev        float64   `json:"stdDev"`
+	Mean          float64   `json:"mean"`
+	Min           float64   `json:"min"`
+	Pct25         float64   `json:"pct25"`
+	Pct50         float64   `json:"pct50"`
+	Pct75         float64   `json:"pct75"`
+	Pct90         float64   `json:"pct90"`
+	Max           float64   `json:"max"`
+}
+
+// add a single result to testTimes
+func (driver *BenchmarkResult) addResult(individualResult testing.BenchmarkResult) {
+	//TODO add assert for single test
+	driver.TestsRun = driver.TestsRun + individualResult.N
+	driver.testTimes = append(driver.testTimes, individualResult.T.Seconds())
+}
+
+// calculate statistical metrics for all testTimes
+func (results *BenchmarkResult) updateStats() {
+	//TODO add assert for single test
+	results.StdDev, _ = stats.StandardDeviation(results.testTimes)
```

Comment:
Just in case, I don't want to throw away the errors here.

Reply:
I will likely refactor this to accumulate errors rather than wrapping every call in the usual error check.

Reply:
I've just pushed a commit that uses github.com/hashicorp/go-multierror to accumulate errors here and return them.